Speed up `roots` when `multiplicities=False` for polynomials over finite fields

c0rydoras opened the pull request:
Speed up `f.roots` when `multiplicities=False` for univariate polynomials over finite fields

- as seen in this blog post: https://adib.au/2025/lance-hard/#speedup-by-using-gcd
- also mentioned here #35556

```sage
sage: p = random_prime(2^200)
sage: F = GF(p)
sage: R.<x> = F[]
sage: f = R.random_element(degree=20000)
sage: %time f.roots(multiplicities=False) # would hang / be very slow before this change
CPU times: user 5.24 s, sys: 16.4 ms, total: 5.26 s
Wall time: 5.26 s
[205828247017582431219769663595709973936079117296023914605767] # just some roots, will ofc be different every run :)
```


### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

c0rydoras committed:
faster roots when multiplicities=False for polynomials over finite fields

user202729 commented on src/sage/rings/finite_rings/integer_mod_ring.py:
```diff
@@ -1935,7 +1935,11 @@ def _roots_univariate_polynomial(self, f, ring=None, multiplicities=True, algori
+            x = R.gen()
+            p = R.modulus()
+            g = pow(x, p, f) - x
+            return [root[0] for root in f.gcd(g).roots()]
```
```suggestion
            g = f.gcd(g)
            return g._roots_from_factorization(g.factor(), False)
```
makes it clearer what gets called later, and avoid the need to add the multiplicity only to remove it later.

c0rydoras replied:
makes sense :)

c0rydoras committed:
add test for faster roots

c0rydoras committed:
Speed up `roots` for all finite fields when `multiplicities=False`

- Add `_roots_univariate_polynomial` to the finite field base class
- `IntegerModRing`s will also use this method to speed up root finding for its polynomials

user202729 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -2120,6 +2120,84 @@ cdef class FiniteField(Field):
+        - ``kwargs`` - ignored
+
+        TESTS::
+            We can take the roots of a polynomial defined over a finite field
```
this documentation is incorrectly formatted (actually this wouldn't show up in the documentation anyway, but you should format it correctly regardless)

Search for `TESTS:` in the code base for examples.

user202729 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -2138,23 +2138,27 @@ cdef class FiniteField(Field):
         - ``kwargs`` - ignored
 
         TESTS::
```
```suggestion
        TESTS:
```

c0rydoras replied:
most of the code does `TESTS::`

```bash
$ rg 'TESTS:[^:]' | wc -l
9
```

```bash
 rg 'TESTS::' | wc -l
11056
```

user202729 replied:
read RST documentation. Even though `TESTS:` is a Sage magic, the whole docstring must be valid.

So, use `TESTS::` if and only if the next paragraph is a code block indented one level more than the rest. If the next paragraph is not a code block, use one colon.

user202729 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -2138,23 +2138,27 @@ cdef class FiniteField(Field):
+
+        We can take the roots of a polynomial defined over a finite field
+
                 sage: set_random_seed(31337)
```
only add 4 spaces of indentation relative to the line above.

c0rydoras replied:
i just adapted what i saw in `src/sage/rings/finite_rings/integer_mod_ring.py`, is there some sort of style guide for this?

user202729 replied:
it's just RST. See also sage coding_basics docs page.

user202729 replied:
> only add 4 spaces of indentation relative to the line above.

this comment is still unaddressed.

user202729 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -2138,23 +2138,27 @@ cdef class FiniteField(Field):
         TESTS::
-            We can take the roots of a polynomial defined over a finite field
+
+        We can take the roots of a polynomial defined over a finite field
```
```suggestion
        We can take the roots of a polynomial defined over a finite field::
```

user202729 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -2120,6 +2120,89 @@ cdef class FiniteField(Field):
+          multiplicity. If ``False``, just return the unique roots, with no
+          information about multiplicities.
+
+        - ``kwargs`` - ignored
```
doesn't look right. I suggest changing `**kwargs` to `algorithm=None`, then in the body, `if algorithm is not None: raise NotImplementedError`, which allows `algorithm=pari` to work automatically.

c0rydoras committed:
fix formatting in docstring && drop kwargs

user202729 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -2120,6 +2120,91 @@ cdef class FiniteField(Field):
+
+        - ``f`` - a polynomial defined over this field
+
+        - ``ring`` - the ring to find roots in.
```
```suggestion
        - ``f`` -- a polynomial defined over this field

        - ``ring`` -- the ring to find roots in
```

and similar below (no full stop if there's only one noun phrase instead of full sentences)

c0rydoras committed:
fix test

c0rydoras committed:
raise NotImplementedError if algorithm != None

c0rydoras committed:
fix ordering mismatch in test

c0rydoras committed:
cast roots back to original ring

c0rydoras committed:
let polynomial_element.pyx's roots handle errors caused by invalid rings

user202729 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -2191,7 +2191,7 @@ cdef class FiniteField(Field):
                     return f._roots_in_subring(L, multiplicities)
                 else:
-                    raise
+                    raise NotImplementedError
```
what's the rationale behind this? (also how do you get the `L.is_exact() and L.is_subring(K)`?)

c0rydoras replied:
from the `roots` in `polynomial_element.pyx`, it only catches `NotImplementedError`s

but we can just let it handle all the cases where `K != L` (https://github.com/sagemath/sage/pull/40494/commits/653ec9fc804a9348410f8d2d98ab89ba7c0be6da)

c0rydoras committed:
clean up copy pasta from polynomial_element.pyx's roots

c0rydoras committed:
fix whitespaces in docstring
